File: sdk/python/vella/pqc/verify_cli.py
"""Historical v3 verification CLI, with explicit historical profile and operator trust."""
import json
import os
import stat
import sys
from typing import Any

from . import bounded_json as bounded
from . import proof


def integer(value: Any) -> bool:
    return type(value) in (int, float) and int(value) == value


def read(path: str, maximum: int, trusted: bool = False) -> Any:
    descriptor = os.open(path, os.O_RDONLY | os.O_NOFOLLOW)
    with os.fdopen(descriptor, 'r', encoding='utf8', errors='strict') as stream:
        info = os.fstat(stream.fileno())
        if not stat.S_ISREG(info.st_mode) or info.st_size > maximum or (trusted and (info.st_uid != os.getuid() or info.st_mode & 0o022)):
            raise ValueError('unsafe or oversized file')
        return bounded.parse(stream.read(maximum + 1), maximum)



def main() -> None:
    try:
        if len(sys.argv) != 5 or sys.argv[4] != proof.SUITE:
            raise ValueError('explicit hybrid profile required')
        bundle = read(sys.argv[1], 2 * 1024 * 1024)
        trust = read(sys.argv[2], 1024 * 1024, True)
        bounded.assert_record_json(trust)
        if type(trust) is not dict or set(trust) != {'kind', 'revision', 'activeId', 'history'} or trust['kind'] != 'vella_local_hybrid_keys_v1' or not integer(trust['revision']) or trust['revision'] < 0 or type(trust['history']) is not list or len(trust['history']) > 256:
            raise ValueError('invalid public trust registry')
        entries = [entry for entry in trust['history'] if entry['id'] == sys.argv[3]]
        if len(entries) != 1 or entries[0]['status'] not in ('active', 'retired', 'revoked'):
            raise ValueError('unrecognized historical key')
        entry = entries[0]
        result = proof.verify(bundle, entry['publicKeys'], sys.argv[4])
        if result['ok']:
            evidence = result['authenticated']['evidence']
            if evidence is not None and 'signing_key' in evidence:
                context = evidence['signing_key']
                if type(context) is not dict or set(context) != {'key_set_id', 'key_revision'} or context['key_set_id'] != entry['id'] or not integer(context['key_revision']) or not 1 <= context['key_revision'] <= trust['revision']:
                    raise ValueError('historical trust binding mismatch')
        result.update(keySetId=entry['id'], keyStatus=entry['status'], trustRevision=trust['revision'], trustedForNewExecution=False)
        result['warnings'].append('Reports the supplied trust snapshot; no trusted signing time or pre-compromise assurance')
    except Exception:  # noqa: BLE001 - reject malformed/untrusted input at the CLI boundary
        result = {'ok': False, 'errors': ['E_PQ_VERIFY_INPUT'], 'warnings': [], 'trustedForNewExecution': False}
    print(json.dumps(result, ensure_ascii=False))
    sys.exit(0 if result['ok'] else 1)


if __name__ == '__main__':
    main()

File: sdk/node/pqc/verify-cli.mjs
#!/usr/bin/env node
// v3 archive CLI. Explicit historical verification, never an execution token.
import fs from 'node:fs';
import {TextDecoder} from 'node:util';
import bounded from './bounded-json.cjs';
import proof from './proof.cjs';
import {matchesArchiveContext} from './archive-trust.cjs';
async function read(path, max, trusted = false) {
  const file = await fs.promises.open(path, fs.constants.O_RDONLY | fs.constants.O_NOFOLLOW);
  try {
    const stat = await file.stat();
    if (!stat.isFile() || stat.size > max || (trusted && (stat.uid !== process.getuid() || (stat.mode & 0o022)))) throw new Error('unsafe or oversized file');
    const bytes = Buffer.alloc(max + 1);
    let length = 0;
    while (length < bytes.length) {
      const {bytesRead} = await file.read(bytes, length, bytes.length - length, null);
      if (!bytesRead) break;
      length += bytesRead;
    }
    if (length > max) throw new Error('file grew beyond limit');
    return bounded.parse(new TextDecoder('utf8', {fatal: true, ignoreBOM: true}).decode(bytes.subarray(0, length)), max);
  } finally { await file.close(); }
}
try {
  const [path, trustPath, id, suite, ...extra] = process.argv.slice(2);
  if (!path || !trustPath || !id || suite !== proof.SUITE || extra.length) throw new Error('explicit proof, public trust, key-set ID and required hybrid suite needed');
  const bundle = await read(path, 2 * 1024 * 1024), trust = await read(trustPath, 1024 * 1024, true);
  bounded.assertRecordJson(trust);
  if (trust.kind !== 'vella_local_hybrid_keys_v1' || !Number.isSafeInteger(trust.revision) || trust.revision < 0 || !Array.isArray(trust.history) || trust.history.length > 256 || !Object.hasOwn(trust, 'activeId') || Object.keys(trust).length !== 4) throw new Error('invalid public trust registry');
  const matches = trust.history.filter(entry => entry.id === id);
  if (matches.length !== 1 || !['active', 'retired', 'revoked'].includes(matches[0].status)) throw new Error('unrecognized historical key');
  const entry = matches[0], result = proof.verify(bundle, entry.publicKeys, suite);
  if (result.ok && !matchesArchiveContext(result.authenticated, id, trust.revision)) throw new Error('historical trust binding mismatch');
  const output = {...result, keySetId: id, keyStatus: entry.status, trustRevision: trust.revision, trustedForNewExecution: false,
    warnings: [...result.warnings, 'Reports the supplied trust snapshot; no trusted signing time or pre-compromise assurance']};
  process.stdout.write(JSON.stringify(output) + '\n'); process.exitCode = result.ok ? 0 : 1;
} catch {
  process.stdout.write(JSON.stringify({ok: false, errors: ['E_PQ_VERIFY_INPUT'], warnings: [], trustedForNewExecution: false}) + '\n'); process.exitCode = 1;
}
